### PR TITLE
Drop Python 2, implement Python 3, replace virtualenv with native python 3 implementation, drop EPEL parameter

### LINKED ISSUE #287: lateset on ubuntu 18.04
since puppetboard dropped support python 2.7 ubuntu 18.04 started failing. as ubuntu 18.04's default is 2.7 for apache::mod::wsgi. this would also mean by default ubuntu 14.04 isn't supported anymore. (metadata.json file)

on ubuntu 18.04 change
```python
  class { 'apache::mod::wsgi': }
```
-->
```puppet
  class { 'apache::mod::wsgi':
    package_name => "libapache2-mod-wsgi-py3",
    mod_path         => "/usr/lib/apache2/modules/mod_wsgi.so",
  }
```

Documentation is heavily RHEL focused and doesn't explain this part well.

can I suggest a variable called $manage_mod_wsgi  and set some sane defaults for these versions?

if so I can push a PR if needed

### PR DESCRIPTION
**Pull request by waipeng**
#### Pull Request (PR) description

This PR attempts to fix tests that have been failing and blocking all PR from being merged.

To support python3, we have to introduce a few new variables to support
this creation of a py3 virtualenv, and also to set wsgi python-home [1]
it. Just using activate_this.py didn't work.

In test, we also choose to use a separate virtualenv
`virtenv3-puppetboard` because `virtenv-puppetboard' will be created as
a py2 virtualenv by the other tests.

[1]: https://modwsgi.readthedocs.io/en/develop/user-guides/virtual-environments.html

#### This Pull Request (PR) fixes the following issues
Fixes #287

### CODE REVIEW

File: .travis.yml
Comment (bastelfreak) [inline .travis.yml:27]: hey. we still support ubuntu16. instead of replacing ubuntu16 with 18, can you add ubuntu18? Please also add ubuntu18 to the .sync.yml file

File: manifests/init.pp
Comment (bastelfreak) [inline manifests/init.pp:202]: instead of setting the static value in params.pp, please directly add it here. When we in the future switch to puppet-strings, puppet-strings will be able to read the value from init.pp. Also I think we can switch the datatype to Stdlib::Absolutepath?
```suggestion
  Stdlib::Absolutepath $virtualenv_dir                        = $puppetboard::params::virtualenv_dir,
```

File: spec/acceptance/class_spec.rb
Comment (bastelfreak) [inline spec/acceptance/class_spec.rb:48]: can you also document this in the README.md?

File: manifests/init.pp
Comment (bastelfreak) [inline manifests/init.pp:130]: can you convert this parameter into puppet-strings style? In the future we want/need to convert all existing params as well.

File: spec/acceptance/class_spec.rb
Comment (bastelfreak) [inline spec/acceptance/class_spec.rb:60]: I really think 3 should be the default, but that maybe should be changed in the future/another PR.

**Comment by bastelfreak** [comment]
thanks for all the work! I already fixed all the linter fails in another PR: https://github.com/voxpupuli/puppet-puppetboard/pull/288

I'm not sure what's the best way to bring both PRs together. Do you want to cherry-pick my commits into your branch? Otherwise I can cherry-pick your changes into mine as soon as the acceptance tests pass.

**Comment by waipeng** [comment]
Thanks for the review!

Currently I'm unable to get the CentOS tests passing. This is due to this fix installs puppetdb so that puppetboard can talk to it. Without a working puppetdb, puppetboard throws a 500 errors. You will notice most of the other tests don't test for connection, maybe this is the case.

On a dockerized CentOS I am unable to get systemd to work properly to start puppetdb properly, it errors out with

```
systemd[1]: New main PID <PID> does not belong to service, and PID file is not owned by root. Refusing.
```

There are two choices:

1) Continue trying fix CentOS and docker and systemd, or wait for someone to jump in and help
2) Change the test to not test for connection, but test for correctness in the generated configs (not as good)

Do you have any ideas?

**Comment by bastelfreak** [comment]
in the past we had strange issues with centos and systemd in docker. using an older version of the container fixed that. Can you use `centos7-64{image=centos:7.6.1810}'` instead of `centos7-64` in the `.travis.yml` and check if the tests pass afterwards?

**Comment by bastelfreak** [comment]
*crosses fingers for the centos tests*

**Comment by waipeng** [comment]
> thanks for all the work! I already fixed all the linter fails in another PR: #288
> 
I noticed the modulesync change removed `BEAKER_HYPERVISOR=docker` so I'm going to try to see if that passes.

**Comment by bastelfreak** [comment]
you can probably let the linter fix the new violations with `bundle exec rake lint_fix`

**Comment by bastelfreak** [comment]
> I noticed the modulesync change removed `BEAKER_HYPERVISOR=docker` so I'm going to try to see if that passes.

yes, that's now set by a gem we require, voxpupuli-acceptance: https://github.com/voxpupuli/voxpupuli-acceptance/blob/master/lib/voxpupuli/acceptance/spec_helper_acceptance.rb#L5

File: manifests/apache/vhost.pp
Comment (ekohl) [inline manifests/apache/vhost.pp:116]: This looks like an unrelated bugfix to modulesync. I think you can also use `+`
```suggestion
  $wsgi_daemon_process_options_default = {
    threads => $threads,
    group   => $group,
    user    => $user,
  } + $wsgi_daemon_process_options_extra
```

I'll leave it up to you to decide if it's more or less readable.

**Review by ekohl — COMMENTED** [review]
I opened https://github.com/voxpupuli/puppet-puppetboard/pull/293 to convert to voxpupuli-acceptance. Entirely untested, but that's probably a good base. If that works, I'd suggest to rebase on that. Also keep the PR focused and not include a full modulesync for easier review.

**Comment by waipeng** [comment]
looks like both ubuntu1604-64 and centos7-64{image=centos:7.6.1810} have problems starting services via systemd. Any ideas how to fix them?

**Comment by bastelfreak** [comment]
I will rebase this on our master

**Comment by bastelfreak** [comment]
*crosses fingers*

**Comment by bastelfreak** [comment]
Hi. i did a lot of debugging to figure out why the virtualenv command isnt found during the tests. My internet just died and I will continue tomorrow

**Comment by waipeng** [comment]
So I've debugged a few different docker images

- **CentOS 7** - systemd 219 - Fail at starting puppetdb
- **Ubuntu 1604** - systemd 229 - Fail at starting postgresql
- **Ubuntu 1804** - systemd 237 - Pass
- **Ubuntu 2004** - systemd 245 - Pass starting puppetdb, but fail other tests (needs work)

I think at this point in time we are debugging systemd/docker issues, not puppetboard/puppet-puppetboard issues.

I'm not sure what is the best way forward. Hoping someone with systemd and docker experience can help?

**Comment by ekohl** [comment]
I would start with (at least locally) testing with vagrant_libvirt. Real VMs at least allow you to exclude failures due to Docker. There's https://github.com/moby/moby/issues/38749 which causes a lot of these issues. Sadly, it's pretty much the only CI we have. I have thought about using podman (which is supposed to be a lot more friendly to systemd inside container) but haven't gotten around to playing with that.

**Comment by bastelfreak** [comment]
I currently use the following command to test on centos7 with virtualbox/vagrant:
```
PUPPET_INSTALL_TYPE=agent BEAKER_destroy=no BEAKER_debug=true BEAKER_PUPPET_COLLECTION=puppet6 BEAKER_setfile=centos7-64{hypervisor=vagrant} BEAKER_TESTMODE=apply bundle exec rake beaker
```

this currently fails with:
```
       	Error: Could not find command 'virtualenv'
```

which should be virtualenv-3. I assume this is a bug in the puppet-python module. I'm currently debugging this and also considering, at least for testing, to switch to pyvenv instead. We already use that in two other modules with python3 on centos 7 so we know it's working.

**Comment by vchepkov** [comment]
virtualenv should be retired, since python3.6 it's no longer needed

**Comment by bastelfreak** [comment]
so what's the recommended way to setp python + packages in a dedicated directory? pyvenv? or calling python directly?  We have https://github.com/voxpupuli/puppet-python/blob/master/manifests/pyvenv.pp and https://github.com/voxpupuli/puppet-python/blob/master/manifests/virtualenv.pp

**Comment by vchepkov** [comment]
pyenv is also deprecated. comment even say so https://github.com/voxpupuli/puppet-python/blob/master/manifests/pyvenv.pp#L58
`python3 -m venv` is the proper way nowadays

**Comment by bastelfreak** [comment]
as a start I added an acceptance test to puppet-python to ensure managing python 3 works on all platforms: https://github.com/voxpupuli/puppet-python/pull/572

**Comment by bastelfreak** [comment]
@vchepkov ah sweet, so calling python::pyvenv which will directly call `python3 -m venv` is the correct way going forward. I will ensure we've proper tests in puppet-python and afterwards update this PR.

**Comment by bastelfreak** [comment]
hey people. so good and bad news here:
* the puppet code works now with centos7 in vagrant/virtualbox and creates a clean python3 environment 
* apache on CentOS 7 has a wsgi module but that's complied against Python2.7 and I run into this error: https://stackoverflow.com/questions/41005030/django-apache-and-virtualenv-importerror-no-module-named-site

I'm not 100% sure how we should proceed here. I think the whole setup would be way easier if puppetboard provided an internal webserver and apache just proxies the requests. The python versions are different on every OS we support which makes the maintenance really painful.

**Comment by ekohl** [comment]
When using a reverse proxy setup, there are additional considerations. For example, header security becomes harder. I would suggest to drop EL7 and move to EL8 since in the short term it's just easier. Python 3 on EL7 was a late addition and EL8 embraces Python 3 as the default.

**Comment by bastelfreak** [comment]
short update here: I managed to get this working properly with python3 on ubuntu18 with vagrant. I will cleanup my local code and check that the other operating systems succeed as well

File: templates/wsgi.py.epp
Comment (ekohl) [inline templates/wsgi.py.epp:11]: Perhaps it can be a static file rather than a template:

```suggestion
me = os.path.dirname(os.path.abspath(__file__))

os.environ['PUPPETBOARD_SETTINGS'] = os.path.join(me, 'settings.py')
```

Untested, but if I read the code correct, this is equivalent

**Comment by waipeng** [comment]
Hi @bastelfreak, I believe I was working on the same thing at the same time! I too got the tests working for both 16.04 and 18.04.

I noticed that puppet-python3 doesn't do any python3 acceptance test unless it is Ubuntu 18.04. This might be why we are running into so many errors running python3 virtualenv?

**Comment by bastelfreak** [comment]
I resolved all python3 errors :) just one job fails, on debian 10 with puppet 5. the error seems unrelated to python. yes the testing in the python module could be heavily improved :(

**Comment by waipeng** [comment]
Yes I saw! Good work! I'll leave you to work on this (I felt like I wasted 3 hours fixing up python3 errors at the same time as you). I will take a look at puppet-python to see I can improve anything.

Feel free to ping me if you need my help, but I think you're doing pretty well :)

**Comment by waipeng** [comment]
@bastelfreak , Debian 10 + Puppet 5 is failing because there is no puppetlabs package for puppetdb. Puppet5 supported versions are Debian 7, 8, 9

https://puppet.com/docs/puppet/5.5/puppet_platform.html

I have further verified by checking the repo http://apt.puppetlabs.com/dists/buster/puppet5/

**Comment by bastelfreak** [comment]
thanks for the fix! I will verify the suggestion from @ekohl and update the epp template. I think afterwards this can get a final review and we merge it *crosses fingers*

File: manifests/apache/vhost.pp
Comment (sebastianrakel) [inline manifests/apache/vhost.pp:115]: There is a basedir parameter which is documented with " Base directory where to build puppetboard vcsrepo and python virtualenv." so should we use this parameter here?

Comment (sebastianrakel) [reply manifests/apache/vhost.pp:115]: And the init.pp had a virtualenv_dir parameter

File: templates/wsgi.py.epp
Comment (bastelfreak) [reply templates/wsgi.py.epp]: tested this locally and applied the patch, thanks!

**Comment by bastelfreak** [comment]
to get a proper changelog, I created some issues and linked them to this PR. I will do some further README.md cleanup now.

**Review by sebastianrakel — COMMENTED** [review]
I mentioned one point, this should be resolved. the rest is okay

File: manifests/apache/vhost.pp
Comment (bastelfreak) [reply manifests/apache/vhost.pp]: thanks, fixed it

**Comment by bastelfreak** [comment]
hey people. I think this is good to go now. I linked multiple issues to this PR which ensures that we get a proper changelog. The parameter documentation is still a mess. I would like to rework it to puppet-strings and remove duplicate / add missing docs, but in an upcoming PR. The diffe here is already pretty huge. Also I want to rebase the modulesync PR as soon as this one is in.

**Review by sebastianrakel — APPROVED** [review]

**Review by mmerfort — APPROVED** [review]

File: manifests/apache/vhost.pp
Comment (ekohl) [inline manifests/apache/vhost.pp:84]: Technically I think `Stdlib::Unixpath` is more correct. It shouldn't accept Windows paths since it's the HTTP path which behaves the same as a unix path.

File: spec/acceptance/class_spec.rb
Comment (ekohl) [inline spec/acceptance/class_spec.rb:29]:
```suggestion
      if $facts['os']['family'] == 'Debian' {
        $mod_path = $facts['os']['release']['major'] ? {
          '18.04' => "/usr/lib/apache2/modules/mod_wsgi.so-3.6",
          '16.04' => "/usr/lib/apache2/modules/mod_wsgi.so-3.5",
          '10'    => "/usr/lib/apache2/modules/mod_wsgi.so-3.7",
          '9'     => "/usr/lib/apache2/modules/mod_wsgi.so-3.5",
          default => undef,
        }
        $wsgi = {
          'package_name' => "libapache2-mod-wsgi-py3",
          'mod_path'     => $mod_path,
        }
      } else {
        $wsgi = {}
      }
```

Also wondering if you can reuse `$puppetboard::python_version` here for `$mod_path`

File: templates/wsgi.py.epp
Comment (ekohl) [inline templates/wsgi.py.epp:1]: There are no variables left. Please move it to `files/wsgi.py` and use `file('puppetboard/wsgi.py')`

File: spec/acceptance/class_spec.rb
Comment (bastelfreak) [reply spec/acceptance/class_spec.rb:29]: good catch. I'm currently testing this locally:

```diff
diff --git a/spec/acceptance/class_spec.rb b/spec/acceptance/class_spec.rb
index a2fd425..4623e6c 100644
--- a/spec/acceptance/class_spec.rb
+++ b/spec/acceptance/class_spec.rb
@@ -16,16 +16,9 @@ describe 'puppetboard class' do
         default_vhost => false,
         purge_configs => true,
       }
-      if $facts['os']['release']['major'] == '20.04' {
-        $wsgi = {package_name => "libapache2-mod-wsgi-py3", mod_path => "/usr/lib/apache2/modules/mod_wsgi.so-3.8",}
-      } elsif $facts['os']['release']['major'] == '18.04' {
-        $wsgi = {package_name => "libapache2-mod-wsgi-py3", mod_path => "/usr/lib/apache2/modules/mod_wsgi.so-3.6",}
-      } elsif $facts['os']['release']['major'] == '16.04' {
-        $wsgi = {package_name => "libapache2-mod-wsgi-py3", mod_path => "/usr/lib/apache2/modules/mod_wsgi.so-3.5",}
-      } elsif $facts['os']['release']['major'] == '10' {
-        $wsgi = {package_name => "libapache2-mod-wsgi-py3", mod_path => "/usr/lib/apache2/modules/mod_wsgi.so-3.7",}
-      } elsif $facts['os']['release']['major'] == '9' {
-        $wsgi = {package_name => "libapache2-mod-wsgi-py3", mod_path => "/usr/lib/apache2/modules/mod_wsgi.so-3.5",}
+      $wsgi = $facts['os']['family'] ? {
+        'Debian' => {package_name => "libapache2-mod-wsgi-py3", mod_path => "/usr/lib/apache2/modules/mod_wsgi.so-${puppetboard::python_version}"},
+        default  => {},
       }
       class { 'apache::mod::wsgi':
         * => $wsgi,

```

Comment (bastelfreak) [reply spec/acceptance/class_spec.rb]: I updated this to rely on the nonversioned .so file. We canot easily access puppetboard::python_version. The .so file could be provided by other packages as well, but they always conflict with libapache2-mod-wsgi-py3. If libapache2-mod-wsgi-py3 got installed we know that the .so file is the correct one.

File: manifests/apache/vhost.pp
Comment (bastelfreak) [reply manifests/apache/vhost.pp]: updated it

File: README.md
Comment (bastelfreak) [reply README.md]: cleaned that up with the code from the spec helper

File: README.md
Comment (ekohl) [inline README.md:159]: This looks a lot like what I would expect from a default module path - is `mod_path` really needed?

Comment (bastelfreak) [reply README.md:159]: sadly it is. apache::mod::wsgi demands that mod_path if set if we provide package_name :(

Comment (ekohl) [reply README.md:159]: Too bad. At least we cleaned up the part that was hard to maintain.

**Review by ekohl — APPROVED** [review]
I would change the title of the PR for the changelog. It ended up also improving docs, add a parameter for daemon opts, use `python::pyvenv`.

**Comment by waipeng** [comment]
Thanks @bastelfreak, @ekohl and all for your help!

Can we have a version released for this please? We are keen to use it.

**Comment by bastelfreak** [comment]
i want to get https://github.com/voxpupuli/puppet-puppetboard/pull/302 in before we do a release

**Pull request merged**
